Stomp server

Hroy02 opened the pull request:
Tell me if anything should have more comments or isn't understandable

Hroy02 committed:
Temp Commit

Hroy02 committed:
Temp Commit

Hroy02 committed:
Remove the newsfeed

Hroy02 committed:
Add .py to gitignore for comfort of usage

Hroy02 committed:
Add server refactoring

Hroy02 committed:
Add libraryProtocol

Hroy02 committed:
Refactor reactor

Hroy02 committed:
Add Frame methods

Hroy02 committed:
Fix LibraryProtocol

Hroy02 committed:
Fix ConnectionsImpl

Hroy02 committed:
Add Comments and some formatting

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/ConnectionsImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+    private ConcurrentHashMap<Integer, ConnectionHandler<Frame>> _connections;
+    // SimpleEntry used as a Pair object
+    private ConcurrentHashMap<String, List<AbstractMap.SimpleEntry<ConnectionHandler<Frame>, Integer>>> _channels;
```
I think this line is too complicated. add explanation/split it

Hroy02 replied:
Not sure how to make it simpler, will add a more elaborate comment

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/ConnectionsImpl.java:
```diff
@@ -0,0 +1,66 @@
+    @Override
+    public void send(String channel, Frame msg) {
+        for(AbstractMap.SimpleEntry<ConnectionHandler<Frame>, Integer> sub : _channels.get(channel)) {
+            sub.getKey().send(addSubscriptionId(msg, sub.getValue())); // Send the message with the subscription header
```
is Connetction handler has a .send function? why not use this.send(int, Frame)?

Hroy02 replied:
a. ConnectionHandler has a send function, thats how this.send(int, Frame) works.
b. you are right I should use it. I will change it

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/ConnectionsImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void subscribe(String channel, int connectionId, int subscriptionId) {
+        synchronized (this) {
```
if you synchronize on this use the syntactic sugar. it is the only place you synchronize. it it good? are you sure you need to synchronize on the whole object?

Hroy02 replied:
I don't want to sync the whole function, just this line, So the syntactic sugar isn't relevant.
This is the only place I synchronize so it will only block two threads from creating this Array at the same time.
I also added another check now so that I will synchronize less. Generally this sync should only happen once for every topic

IdanTo1 replied:
my bad, didn't notice where the `}` goes.. so you're good (:

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/Frame.java:
```diff
@@ -0,0 +1,89 @@
+    public String toString() {
+        StringBuilder s = new StringBuilder(_command.name() + "\n");
+        for (Map.Entry<String, String> header : _headers.entrySet()) {
+            s.append(header.getKey()).append(":").append(header.getValue());
```
I don't like this `.append` concatenation.. isn't `+ ":" + header...` possible? think of a better way

Hroy02 replied:
I actually started out using operator+ but IntelliJ offered to change to this. I can easily change it back, but this seemed like the more proper way to do this

IdanTo1 replied:
in my opinion, it is kind of messy... i would prefer the operator+
in the and, your choice

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/Frame.java:
```diff
@@ -0,0 +1,89 @@
+        for (Map.Entry<String, String> header : _headers.entrySet()) {
+            s.append(header.getKey()).append(":").append(header.getValue());
+        }
+        s.append(_body == null ? "" : _body);
```
no `\n` needed?

Hroy02 replied:
Very needed, nice catch!

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/Frame.java:
```diff
@@ -0,0 +1,89 @@
+    public Frame clone() {
+        Frame f = new Frame();
+        f.setCommand(_command);
+        f.addBody(_body);
```
use `Frame(command, body)` instead

Hroy02 replied:
Done

IdanTo1 reviewed (changes requested):
this will be a good start :)

Hroy02 committed:
Small fix needed

Hroy02 committed:
Fix sending the Frame

Hroy02 committed:
Multiple

1. Add if before sync to sync less
2. Save connectionId instead of the entire ConnectionHandler to avoid code duplication

Hroy02 committed:
Fix Frame.toString() to add the needed \n's

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+     */
+    private Frame parseMessage(String msg) {
+        int delim = 0, prevDelim = 0;
+        int msgPart = 1;
```
why are you starting from 1 and not 0?

Hroy02 replied:
Absolutely no reason at all. Comfort I suppose

IdanTo1 replied:
so change it to 0, and change the magical numbers

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+        do {
+            prevDelim = delim;
+            delim = msg.indexOf("\n", prevDelim+1); // Find the end of the next line
+            if (msgPart == 1) { // The command
```
save it as a const, don't use magic numbers 😄

Hroy02 replied:
Done

IdanTo1 replied:
meant the msgPart == 1 :sweat_smile:

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+            }
+            else if (msgPart == 2) { // headers
+                int mid = msg.indexOf(":", prevDelim);
+                if(mid == -1) {
```
add a comment, no headers found

Hroy02 replied:
Done

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+
+
+        } while (delim < msg.length());
+        return null; // Impossible
```
maybe return parsed here, and in (msgPart == 3) add delim = msg.length()
or even better, while (msgPart <3) and after while loop add body

Hroy02 replied:
Done

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+                parsed.addHeader(msg.substring(prevDelim + 1, mid), msg.substring(mid + 1, delim));
+            } // Body and return
+            else if (msgPart == 3) {
+                // -2 so we don't include the \n or the \0
```
?

Hroy02 replied:
My bad. Changed the code and didn't fix the comment.

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -0,0 +1,82 @@
+        }
+
+        pushByte(nextByte);
+        return null; //not a line yet
```
not a STOMP message yet

Hroy02 replied:
Right!

IdanTo1 commented on src/main/java/bgu/spl/net/srv/BaseServer.java:
```diff
@@ -10,19 +12,24 @@
         this.encdecFactory = encdecFactory;
 		this.sock = null;
+		this.connections = connections;
+		this.currentId = 0;
```
there's some weird indentation ..

Hroy02 replied:
Looks like some tab-space issue, I think it should be fixed now

IdanTo1 reviewed (changes requested):
looks great, no serious changes

Hroy02 committed:
As per review

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -23,28 +24,26 @@ private Frame parseMessage(String msg) {
 
-        } while (delim < msg.length());
-        return null; // Impossible
+        } while (delim < msg.length() && msgPart <= 3);
```
i meant < 3
and than put all the body's logic outside the while loop

Hroy02 committed:
As per review

Hroy02 committed:
Remove magic numbers

IdanTo1 commented on src/main/java/bgu/spl/net/impl/stomp/StompEncDec.java:
```diff
@@ -12,6 +12,9 @@
     private static final char delimiter = '\n';
+    private static final int COMMAND_PART = 0;
+    private static final int HEADERS_PART = 1;
+    private static final int PARTS = 2;
```
body?

Hroy02 committed:
Rename constant
